### ISSUE TITLE
Swift support?

### ISSUE THREAD

**Issue by konrad1977** [joaotavora/eglot#825]
Hi i couldnt find support for swift. It's this something that is planned or easily added?

**Comment by skangas** [comment]
Hi!

In most cases it should be trivial to add to `eglot-server-programs` using something like:

```
(add-to-list 'eglot-server-programs '(foo-mode . ("foo-language-server" "--args")))
```

If you could tell us the mode and the command to start the language server, it would be trivial to add it. Some language servers are worse than others though, and might need particular support code (e.g. Java). I don't know if Swift is one of them.

**Comment by konrad1977** [comment]
Yeah, I got it working. But it doesnt complete-at-point. Nor does lsp-sourcekit maybe something is wrong with apples lsp server? According to Apple it has a lot of things supported that doesnt work. Like rename https://github.com/apple/sourcekit-lsp

There is a command to find lsp-sourcekit on mac. 
$ `xcrun --find sourcekit-lsp`
prints the following : /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/sourcekit-lsp

Let me know if I can be of any assistance.

**Comment by konrad1977** [comment]
Did a comparison with VSCode and SDE (Swift Development) and they seem to work the same. Its probably lsp-sourcekit that doesn't work correctly. I don't get completion of inner variables of structs/classes. Only current file.

Here is a picture comparing VS Code and Emacs and Eglot.
<img width="1231" alt="Skärmavbild 2022-01-28 kl  12 33 40" src="https://user-images.githubusercontent.com/4384932/151540230-1a0f5804-fd73-493d-8fa0-9dbf6aa17ab3.png">
<img width="489" alt="Skärmavbild 2022-01-28 kl  12 33 31" src="https://user-images.githubusercontent.com/4384932/151540242-59ab80c3-3df0-49a2-8e68-81afee8ee03e.png">

**Comment by nemethf** [comment]
>  Its probably lsp-sourcekit that doesn't work correctly. 

I don't follow this discussion closely, but there's seems to be a misunderstanding.  lsp-sourcekit requires lsp-mode and not Eglot.  However, if the problem turns out to be in Eglot, filling out a proper bug report will greatly help the developers of Eglot to work on the issue.

Also, @konrad1977 said he "got it working", but for others to benefit, we need the exact settings of eglot-server-programs.

**Comment by konrad1977** [comment]
I am new to lsp/eglot and emacs in general. But I thought you could got get full syntax-highlighting and code-completion. But maybe that's not the case? 

I meant that the bugs are in https://github.com/apple/sourcekit-lsp and not in lsp-sourcekit (which is now a part of lsp-mode).

This is how I load swift support for eglot. I take no credibility (found it on GitHub)

`  (setq mk-sourcekit-lsp-options '("--sync"))
  (defun mk-sourcekit-lsp-executable ()
	(setq mk-sourcekit-lsp-executable
          (cond ((executable-find "sourcekit-lsp"))
				((equal system-type 'darwin)
				 (cond ((executable-find "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/sourcekit-lsp"))
					   ((executable-find "/usr/local/bin/sourcekit-lsp"))
                       ((executable-find "/Library/Developer/CommandLineTools/usr/bin/sourcekit-lsp"))))
				((equal system-type 'gnu/linux)
				 (cond ((executable-find "/home/linuxbrew/.linuxbrew/bin/sourcekit-lsp"))))
				(t
				 ("sourcekit-lsp")))))

  (defun mk-sourcekit-lsp-command (interactive)
	(append (list (mk-sourcekit-lsp-executable)) mk-sourcekit-lsp-options))

  (use-package eglot
	:config
	(add-to-list 'eglot-server-programs '((swift-mode) . mk-sourcekit-lsp-command)))
`

**Comment by woolsweater** [comment]
There's some unfortunate naming confusion here. `sourcekit-lsp` is Apple's implementation of an LSP server, based on its SourceKit tool. `lsp-sourcekit` is an Emacs package that plugs the Apple LSP server into `lsp-mode` (another Emacs package).

If you are using Eglot, you do not need `lsp-sourcekit`. You just need to tell Eglot where the `sourcekit-lsp` binary is on your system. That's what your snippet does.

In some cases (for iOS development, for example), you might also want to pass some compiler flags, such as the iOS SDK. You can see [this for example in my Emacs config](https://gitlab.com/woolsweater/dotemacs.d/-/blob/main/modules/my-swift-mode.el#L120-183). (Please feel free to use/adapt that code if you wish.)

As for the behavior, Apple's LSP server does not recognize Xcode-based projects. In order to get project-wide completion it is pretty much required to have a Package.swift manifest. (There are suggestions scattered around the web for alternatives. For example, getting a 'compile_commands.json' file for an Xcode project. `sourcekit-lsp` will try use that as a fallback, but I haven't found this or other options to work well at all.)

**Comment by joaotavora** [comment]
Nice explanation . Going to try to convert this into a discussion and mark your post the answer.
